src/main/scala/patmat/Huffman.scala

```diff
@@ -0,0 +1,342 @@
+
+    def makeOrderedLeafListHelper(freqs: List[(Char, Int)]): List[Leaf] = freqs match {
+      case Nil          => Nil
+      case head :: tail => Leaf(head._1, head._2) :: makeOrderedLeafList(tail)
+    }
+
+    makeOrderedLeafListHelper(freqs.sortWith((el1, el2) => el1._2 < el2._2))
+  }
+
+  /**
+    * Checks whether the list `trees` contains only one single code tree.
+    */
+  def singleton(trees: List[CodeTree]): Boolean = trees match {
+    case _ :: Nil => true
+    case _        => false
+  }
+
+  /**
+    * The parameter `trees` of this function is a list of code trees ordered
+    * by ascending weights.
+    *
+    * This function takes the first two elements of the list `trees` and combines
+    * them into a single `Fork` node. This node is then added back into the
+    * remaining elements of `trees` at a position such that the ordering by weights
+    * is preserved.
+    *
+    * If `trees` is a list of less than two elements, that list should be returned
+    * unchanged.
+    */
+  def combine(trees: List[CodeTree]): List[CodeTree] = trees match {
+    case tree1 :: tree2 :: tail =>
+      (makeCodeTree(tree1, tree2) :: tail).sortWith((el1, el2) => weight(el1) < weight(el2))
+    case _ => trees
+  }
+
+  /**
+    * This function will be called in the following way:
+    *
+    * until(singleton, combine)(trees)
+    *
+    * where `trees` is of type `List[CodeTree]`, `singleton` and `combine` refer to
+    * the two functions defined above.
+    *
+    * In such an invocation, `until` should call the two functions until the list of
+    * code trees contains only one single tree, and then return that singleton list.
+    *
+    * Hint: before writing the implementation,
+    *  - start by defining the parameter types such that the above example invocation
+    *    is valid. The parameter types of `until` should match the argument types of
+    *    the example invocation. Also define the return type of the `until` function.
+    *  - try to find sensible parameter names for `xxx`, `yyy` and `zzz`.
+    */
+  @tailrec def until(singleton: List[CodeTree] => Boolean, combine: List[CodeTree] => List[CodeTree])(
+      trees: List[CodeTree]): List[CodeTree] =
+    if (singleton(trees)) trees
+    else until(singleton, combine)(combine(trees))
+
+  /**
+    * This function creates a code tree which is optimal to encode the text `chars`.
+    *
+    * The parameter `chars` is an arbitrary text. This function extracts the character
+    * frequencies from that text and creates a code tree based on them.
+    */
+  def createCodeTree(chars: List[Char]): CodeTree =
+    until(singleton, combine)(makeOrderedLeafList(times(chars))).head
+
+  // Part 3: Decoding
+
+  type Bit = Int
+
+  /**
+    * This function decodes the bit sequence `bits` using the code tree `tree` and returns
+    * the resulting list of characters.
+    */
+  def decode(tree: CodeTree, bits: List[Bit]): List[Char] = tree match {
+    case Leaf(char, _) =>
```

Comment:
You can use guardians here. It will be more readable. 

```
case Leaf(char, _) if bits.isEmpty => char :: Nil
case Leaf(char, _) => char :: decode(tree, bits.tail)
```